File: Moobeez/Classes/Moobeez/Views/StarsView.m
//
//  StarsView.m
//  Moobeez
//
//  Created by Radu Banea on 10/23/13.
//  Copyright (c) 2013 Goggzy. All rights reserved.
//

#import "StarsView.h"
#import "Moobeez.h"

#define MAX_RATING 5

@interface StarsView ()

@property (weak, nonatomic) IBOutlet UIImageView* emptyStarsImageView;
@property (weak, nonatomic) IBOutlet UIImageView* fullStarsImageView;

@end

@implementation StarsView

- (void)awakeFromNib {
    _rating = MAX_RATING;
}

- (IBAction)starsViewDidPan:(id)sender {
    
    UIPanGestureRecognizer* gesture = (UIPanGestureRecognizer*) sender;
    switch (gesture.state) {
        case UIGestureRecognizerStateBegan:
            break;
        case UIGestureRecognizerStateChanged:
        {
            int point = [gesture locationInView:self].x;
            
            point = MIN(point, self.width);
            point = MAX(point, 0);
            
            point = roundf((float) point / (self.width / 10)) * (self.width / 10);
            
            self.ratingInPixels = point;
        }
            break;
        case UIGestureRecognizerStateEnded:
        case UIGestureRecognizerStateCancelled:
        {
            _rating = self.ratingInPixels / (self.emptyStarsImageView.width + self.fullStarsImageView.width) * MAX_RATING;
            self.updateHandler();
        }
            break;
        default:
            break;
    }
    
}

- (void)setRating:(CGFloat)rating {

    if (_rating > 0) {
        self.ratingInPixels = self.fullStarsImageView.width * rating / _rating;
    }
    else {
        self.ratingInPixels = self.emptyStarsImageView.width * rating / MAX_RATING;
    }

    _rating = rating;
}

- (void)setRatingInPixels:(CGFloat)ratingInPixels {
    _ratingInPixels = ratingInPixels;
    
    self.fullStarsImageView.width = ratingInPixels;
    int right = self.emptyStarsImageView.width + self.emptyStarsImageView.x;
    self.emptyStarsImageView.x = self.fullStarsImageView.x + self.fullStarsImageView.width;
    self.emptyStarsImageView.width = right - self.emptyStarsImageView.x;
}

@end

File: Moobeez/Classes/UI/General/Stars View/StarsView.swift
//
//  StarsView.swift
//  Moobeez
//
//  Created by Radu Banea on 11/09/2017.
//  Copyright © 2017 Gogo's Lab. All rights reserved.
//

import UIKit

class StarsView: UIView {

    @IBOutlet var emptyStarsImageView: UIImageView!
    @IBOutlet var fullStarsImageView: UIImageView!
    
    @IBOutlet var widthConstraint: NSLayoutConstraint!

    private static var maxRating: CGFloat = 5
    
    private var panGesture:UIPanGestureRecognizer?
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        panGesture = UIPanGestureRecognizer(target: self, action: #selector(self.handlePan(recognizer:)))
        addGestureRecognizer(panGesture!)
        
    }
    var rating: CGFloat = maxRating {
        willSet(newRating) {
            percent = newRating / StarsView.maxRating
        }
    }
    var ratingInPixels: CGFloat = 0 {
        
        didSet {
            percent = ratingInPixels / width
        }
        
    }
    
    var percent: CGFloat = 0 {
        didSet {
            removeConstraint(widthConstraint)
            widthConstraint = NSLayoutConstraint(item: fullStarsImageView, attribute: NSLayoutConstraint.Attribute.width,
                                                 relatedBy: NSLayoutConstraint.Relation.equal,
                                                 toItem: self, attribute: NSLayoutConstraint.Attribute.width,
                                                 multiplier: percent, constant: 0)
            addConstraint(widthConstraint)
        }
    }

    var updateHandler: EmptyHandler?

    @objc @IBAction func handlePan(recognizer:UIPanGestureRecognizer) {
        
        if recognizer.state == .began {
            
        }
        else if recognizer.state == .changed {
            
            var point:CGFloat = recognizer.location(in: self).x
            
            point = max(min(point, width), 0)
            
            point = round(point / (width / 10)) * (width / 10)
            
            ratingInPixels = point;
            
        }
        else if recognizer.state == .ended || recognizer.state == .cancelled {
            rating = ratingInPixels / width * StarsView.maxRating
            
            if updateHandler != nil {
                updateHandler!()
            }
            
        }
    }
}
